compiler/luci/service/src/Nodes/CircleFullyConnected.cpp

```diff
@@ -36,4 +40,68 @@ luci::CircleNode *CloneNodeLet<CN::DEF>::visit(const luci::CircleFullyConnected
   return cloned;
 }
 
+namespace sinf
+{
+
+loco::TensorShape Algorithm::visit(const luci::CircleFullyConnected *node)
+{
+  auto input_shape = circle_shape(loco::must_cast<CircleNode *>(node->input()));
+  auto weights_shape = circle_shape(loco::must_cast<CircleNode *>(node->weights()));
+
+  loco::TensorShape out_shape;
+
+  // NOTE Some recipes in some repositories are using rank 4 input for FullyConnected.
+  //      Until they are all fixed, disable following assert.
+  // TODO Enable following assert after related fixes are applied
+  // https://github.com/tensorflow/tensorflow/blob/ea33c1e7a25d8025e8ee405ad8ab7be261798d76/tensorflow/lite/kernels/fully_connected.cc#L194
+  // LUCI_ASSERT(input_shape.rank() == 2 || input_shape.rank() == 3,
+  //             "Input rank of FullyConnected should be 2 or 3");
+
+  // https://github.com/tensorflow/tensorflow/blob/ea33c1e7a25d8025e8ee405ad8ab7be261798d76/tensorflow/lite/kernels/fully_connected.cc#L225
+  LUCI_ASSERT(weights_shape.rank() == 2, "Weights of FullyConnected should be 2");
+
+  // https://github.com/tensorflow/tensorflow/blob/ea33c1e7a25d8025e8ee405ad8ab7be261798d76/tensorflow/lite/kernels/fully_connected.cc#L353-L367
+  if (node->keep_num_dims())
+  {
+    out_shape.rank(input_shape.rank());
+    for (uint32_t i = 0; i < input_shape.rank(); ++i)
+      out_shape.dim(i) = input_shape.dim(i);
+    out_shape.dim(out_shape.rank() - 1) = weights_shape.dim(0);
+  }
+  else
+  {
+    bool is_dynamic_shape = false;
+
+    for (uint32_t i = 0; i < input_shape.rank() - 1; i++)
+    {
+      if (not input_shape.dim(i).known())
+      {
+        is_dynamic_shape = true;
+        break;
+      }
+    }
+
+    // The output BatchSize is determined by the input_size,
+    // which is calculated by multiplying dimensions up to rank()-2.
+
+    uint32_t batch_size = 1;
+
+    for (uint32_t i = 0; i < input_shape.rank() - 1; i++)
+    {
+      batch_size *= input_shape.dim(i).value();
+    }
+
+    out_shape.rank(2);
+    if (is_dynamic_shape)
+      out_shape.dim(0).unset();
```

Comment:
plz add some tests for this unknown dim

Reply:
Q) why is first dim set to unknown for `is_dynamic_shape`?

Reply:
> Q) why is first dim set to unknown for `is_dynamic_shape`?

If there is at least one unknown dimension in the input_shape, it is set to 0, resulting in a batch size calculation of 0.

I will add test for that case and include the explanation as a comment.

Reply:
To explain in more detail, Batch size is determined by the input_size by multiplying the value of each dimension. Therefore, if there is at least one dynamic dimension, the batch size may become 0.

* keep_num_dims ==true
[1,?,15] -> [1,?,20] ( weight_shape =[20,15])
* keep_num_dims ==false
[1,?,15] -> [?,20] ( weight_shape =[20,15])